// @prisma/language-server/dist/__test__/quickFix/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertQuickFix = assertQuickFix;
exports.createDiagnosticErrorUnknownType = createDiagnosticErrorUnknownType;
const vitest_1 = require("vitest");
const vscode_languageserver_1 = require("vscode-languageserver");
const Schema_1 = require("../../lib/Schema");
const code_actions_1 = require("../../lib/code-actions");
function assertQuickFix(expected, textDocument, range, diagnostics) {
    const params = {
        textDocument: {
            uri: textDocument.uri,
        },
        context: {
            diagnostics,
        },
        range,
    };
    const quickFixResult = (0, code_actions_1.quickFix)(Schema_1.PrismaSchema.singleFile(textDocument), textDocument, params);
    (0, vitest_1.expect)(quickFixResult.length).toBeGreaterThan(0);
    (0, vitest_1.expect)(quickFixResult).toStrictEqual(expected);
}
function createDiagnosticErrorUnknownType(unknownType, range) {
    return {
        message: `Type "${unknownType}" is neither a built-in type, nor refers to another model, custom type, or enum.`,
        severity: vscode_languageserver_1.DiagnosticSeverity.Error,
        range: range,
    };
}
